lasagne/layers/recurrent.py

```diff
@@ -757,3 +762,306 @@ def step_masked(input_dot_W_n, mask_n, cell_previous, hid_previous,
+            name="W_hid_to_updategate")
+
+        self.b_updategate = self.add_param(
+            b_updategate, (num_units,),
+            name="b_updategate", regularizable=False)
+
+        self.W_in_to_resetgate = self.add_param(
+            W_in_to_resetgate, (num_inputs, num_units),
+            name="W_in_to_resetgate")
+
+        self.W_hid_to_resetgate = self.add_param(
+            W_hid_to_resetgate, (num_units, num_units),
+            name="W_hid_to_resetgate")
+
+        self.b_resetgate = self.add_param(
+            b_resetgate, (num_units,),
+            name="b_resetgate", regularizable=False)
+
+        self.W_in_to_cell = self.add_param(
+            W_in_to_cell, (num_inputs, num_units), name="W_in_to_cell")
+
+        self.W_hid_to_cell = self.add_param(
+            W_hid_to_cell, (num_units, num_units), name="W_hid_to_cell")
+
+        self.b_cell = self.add_param(
+            b_cell, (num_units,), name="b_cell", regularizable=False)
+
+        self.W_in_to_gates = T.concatenate(
+            [self.W_in_to_resetgate, self.W_in_to_updategate,
+             self.W_in_to_cell], axis=1)
+
+        # Same for hidden to gate weight matrices
+        self.W_hid_to_gates = T.concatenate(
+            [self.W_hid_to_resetgate, self.W_hid_to_updategate,
+             self.W_hid_to_cell], axis=1)
+
+        # Stack gate biases into a (4*num_units) vector
+        self.b_gates = T.concatenate(
+            [self.b_resetgate, self.b_updategate,
+             self.b_cell], axis=0)
+
+        # Initialize hidden state
+        if isinstance(hid_init, T.TensorVariable):
+            if hid_init.ndim != 2:
+                raise ValueError("When a tensor hid_init should be a matrix")
+            self.hid_init = hid_init
+        else:
+            self.hid_init = self.add_param(
+                hid_init, (1, self.num_units), name="hid_init",
+                trainable=learn_init, regularizable=False)
+
+    def get_output_shape_for(self, input_shape):
+        return input_shape[0], input_shape[1], self.num_units
+
+    def get_output_for(self, input, mask=None, *args, **kwargs):
+        """
+        Compute this layer's output function given a symbolic input variable
+
+        Parameters
+        ----------
+        input : theano.TensorType
+            Symbolic input variable
+        mask : theano.TensorType
+            Theano variable denoting whether each time step in each
+            sequence in the batch is part of the sequence or not.  If None,
+            then it assumed that all sequences are of the same length.  If
+            not all sequences are of the same length, then it must be
+            supplied as a matrix of shape (n_batch, n_time_steps) where
+            `mask[i, j] = 1` when `j <= (length of sequence i)` and
+            `mask[i, j] = 0` when `j > (length of sequence i)`.
+        """
+
+        # Treat all layers after the first as flattened feature dimensions
```

Comment:
This should say "Treat all dimensions after the second"; this is an error in all of the layers.